# Classical Computing & Digital Logic
---

Before diving into the world of quantum computing, it is helpful to do a quick review of some of the basics of classical computation and communication. This will allow us to gradually introduce some of the most important concepts that make quantum information different from its classical counterpart. 

Even if you understand digital systems fairly well, it is still worth to at least skim through this chapter since we will be discussing how the model of computation can be expanded to incorporate randomness into our systems, and how logic gates can be modified to make operations reversible.

## 1. Binary Numbers

All modern computers and communication systems rely on the use of binary numbers. The fundamental unit of a binary number is the **bit**, which can take only two values: $0$ or $1$, but just like with decimal numbers, we can express larger binary values by concatenating bits together. For example, with 3 bits, we can express up to 8 different binary numbers:

| Binary | Decimal |
| :-: | :-: |
| 000 | 0 |
| 001 | 1 |
| 010 | 2 |
| 011 | 3 |
| 100 | 4 |
| 101 | 5 |
| 110 | 6 |
| 111 | 7 |

In general, $n$ bits allow us to express up to $2^n$ different values. So, an $n$-bit binary number $b$ can be represented as:

$$ b = b_{n-1} ... b_1 b_0 ,$$

where each bit $b_i$ can take the value $0$ or $1$. 

In this notation, the bit most to the left, $b_0$, is known as the least significant bit (LSB). The bit most to the right, $b_{n-1}$, is known as the most significant bit (MSB).

To convert a binary number $b$ to a decimal integer $x$, we simply take each of the individual bits, and multiply them by 2 exponentiated to the bit's significance, and then sum them up:

$$ x =  2^{n-1} b_{n-1} + ... + 2^{1} b_{1} + 2^{0} b_{0}, $$

which we can write more compactly as using big-sum notation:

$$ x = \sum_{i = 0}^{n-1} 2^{i} b_i .$$

So, to convert, for example, the binary value $1101$ to decimal, we get:

$$ 
\begin{aligned}
x &=  2^{3} b_{3} + 2^{2} b_{2} + 2^{1} b_{1} + 2^{0} b_{0}
\\
\\
x &=  8 \times 1 + 4 \times 1 + 2 \times 0 + 1 \times 1
\\
\\
x &=  13
\end{aligned}
$$

Doing this manually is a tedious process. Luckily we can use Python to do this conversion for us. In Python we can represent binary numbers by using the prefix `0b`. These values still get treated like other integers, so if we, for example, print them, the output displays the corresponding decimal value:

In [1]:
x = 0b1101
print(x)

13


Using this notation, we can also perform the same arithmetic operations we use for decimal integers:

In [2]:
x = 0b1011 + 0b1001  # in decimal: 11 + 9 = 20
print(x)

20


Now, to convert a decimal integer $x$ into a binary number $b$, we can find the $i^{\text{th}}$ bit of $b$ using the following expression:

$$ b_i = \left\lfloor\frac{x}{2^i}\right\rfloor \text{mod } 2 . $$

Here, $\lfloor \cdot \rfloor$ is the [floor function](https://en.wikipedia.org/wiki/Floor_and_ceiling_functions), which returns the integer part of its input. Furthermore, $a \text{ mod }m$ represents the [modulo operation](https://en.wikipedia.org/wiki/Modulo), which returns the remainder of diving $a$ by $m$.
Therefore, the binary representation of $x$ is given by:

$$ b = \sum_{i = 0}^{n-1} 10^{i} \left(\left\lfloor\frac{x}{2^i}\right\rfloor \text{mod } 2 \right).$$

So, to convert the decimal integer $13$ into a 4-bit binary number, we would take:

$$ 
\begin{aligned}
b &= 10^{3} \left(\left\lfloor\frac{13}{2^3}\right\rfloor \text{mod } 2 \right) +
     10^{2} \left(\left\lfloor\frac{13}{2^2}\right\rfloor \text{mod } 2 \right) +
     10^{1} \left(\left\lfloor\frac{13}{2^1}\right\rfloor \text{mod } 2 \right) +
     10^{0} \left(\left\lfloor\frac{13}{2^0}\right\rfloor \text{mod } 2 \right)
\\
\\
b &= 10^{3} \left(1 \text{ mod } 2 \right) +
     10^{2} \left(3 \text{ mod } 2 \right) +
     10^{1} \left(6 \text{ mod } 2 \right) +
     10^{0} \left(13 \text{ mod } 2 \right)
\\ 
\\
b &= 1000 \times 1 +
     100 \times 1 +
     10 \times 0 +
     1 \times 1
\\  
\\
b &= 1101
\\
\end{aligned}
$$

One important note here is that, in this example, we picked the exact number of bits to get the correct binary representation for $13$. However, if we would have picked $n$ to be $3$ or less, we would have ended up with the incorrect representation. So, to guarantee we can correctly express a decimal integer (greater than $0$) in binary, we need to make sure to pick $n$ to be:

$$ n \geq \lfloor \log_2 (x) + 1 \rfloor , $$

where again, $\lfloor \cdot \rfloor$ is the [floor function](https://en.wikipedia.org/wiki/Floor_and_ceiling_functions). So, to accurately represent $13$ in binary, we need:

$$ n \geq \lfloor \log_2 (13) + 1 \rfloor = 4 . $$

Again, the concepts above are important to deepen our understanding, but we do not need to go through the pain of calculating every one of these steps when we can simply rely on Python. So, to convert a decimal integer to binary, we can use the `bin()` function:

In [3]:
b = bin(13)
print(b)

0b1101


Now, one thing to note here is that the `bin()` function does not return an integer value, but rather a string of characters that represents our number in binary. For example, as we saw before, representing number $13$ as `0b1101` returns an integer (in Python denoted as `int`):

In [4]:
print(type(0b1101))

<class 'int'>


whereas expressing $13$ in binary using `bin` returns a string (`str`):

In [5]:
print(type(bin(13)))

<class 'str'>


This means that we can't really perform arithmetic operations directly on these objects and get the right result. For example, if we try to "add" two binary strings, Python treats the symbol `+` as a concatenation operator rather than numerical addition:

In [6]:
b = bin(11) + bin(9)  # in decimal: 11 + 9 = 20
print(f'The result in b: {b} is not the correct representation for the number 20: {bin(20)}')

The result in b: 0b10110b1001 is not the correct representation for the number 20: 0b10100


Therefore, we need to be careful when dealing with binary strings and always make sure they are in the right format when we want to perform certain operations. Now, luckily, we can always convert back a binary string into an integer in Python by using the `int()` function:

In [7]:
b = bin(13)  # converts 13 into a binary string
print(f'value in binary: {b}')
x = int(b, 2) # converts the binary string back to an integer
print(f'value in decimal: {x}')

value in binary: 0b1101
value in decimal: 13


Notice that when we did `int(b, 2)`, we not only passed the value we want to convert back to decimal (`b`), but also base that number is expressed in (in this case, `2` for binary). So, if we have binary strings and want to perform certain operations on them, we can always convert them back to decimal integers using this function, and then convert them back to binary strings. 

It is also worth noting that the strings we pass to the `int` function do not necessarily have to have the prefix `0b`. For instance:

In [8]:
b = '1101'
print(int(b,2))

13


This means that we can often drop the `0b` prefix of a binary representation without having to worry about adding it again if we need to convert the value back to decimal. Since strings in Python are indexable, we can easily do this by simply storing the portion of string after the first two characters:

In [9]:
b = bin(13)[2:]
print(b)

1101


Another important aspect of using the `bin()` function is that it always returns a string using the minimum number of bits required to represent the input number. For instance, notice how the number of bits changes below from 1 to 3 as we increase the decimal value:

In [10]:
for i in range(2**3):
    b = bin(i)[2:]
    print(f'{i}: {b}')

0: 0
1: 1
2: 10
3: 11
4: 100
5: 101
6: 110
7: 111


However, we will often want to express different binary numbers using a fix number of bits. For this, we can pad our string with zeros by using the `zfill()` method:

In [11]:
n = 4  # number of bits to use
for i in range(2**3):
    b = bin(i)[2:].zfill(n)
    print(f'{i}: {b}')

0: 0000
1: 0001
2: 0010
3: 0011
4: 0100
5: 0101
6: 0110
7: 0111


Alternatively, we can make use of the first library we will introduce in this textbook: [NumPy](https://numpy.org/). NumPy is a Python library optimized for numerical computation of multi-dimensional arrays and matrices. NumPy also counts with a large number of useful functions, including one to convert decimal numbers to binary strings of an arbitrary length.

Below, we first import NumPy using the alias `np`, and the make use of the function `binary_repr()` which takes as an input the decimal integer we want to convert, and the number of bits to represent it. For example, if we want the number $13$ represented in $5$ bits, we get $01101$:

In [12]:
import numpy as np

b = np.binary_repr(13, 5)
print(b)

01101


Now that we have developed an understanding of how to use binary numbers in Python, we can discuss the types of operations we can perform on them and how are these implemented in the form of digital circuits. 

## 2. Boolean Logic

[Boolean logic](https://en.wikipedia.org/wiki/Boolean_algebra) is the type of algebra used to operate on binary numbers. In Boolean logic, variables can take one of two possible values: **True** or **False**, which are commonly represented with the values we previously assigned for the **bit**: $1$ or $0$, respectively.

There are three basic operations associated with Boolean algebra: [Negation](https://en.wikipedia.org/wiki/Negation) (**NOT**), [Conjunction](https://en.wikipedia.org/wiki/Logical_disjunction) (**OR**), and [Disjunction](https://en.wikipedia.org/wiki/Logical_conjunction) (**AND**).

<br></br>
The **NOT** operation is unary, meaning that it operates only on one variable by negating its value. There are many different symbols used to represent negation, but here we will denote it by using a bar over the variable being negated. A convenient way to express Boolean operations is by the use for Truth tables; the one below corresponds to that of the **NOT** operation:

| $a$ | $\bar{a}$ |
| :-: | :-: |
| $0$ | $1$ |
| $1$ | $0$ |
<br></br>

The **OR** operation is binary (operates on two variables), and results in a **True** statement when either of the variables are **True** (i.e., either of them is equal to $1$). Again, there are many different symbols to represent conjunction, but here we will use $(b \lor a)$ to denote $b$ **OR** $a$:

| $b$ | $a$ | $b \lor a$ |
| :-: | :-: | :-: |
| $0$ | $0$ | $0$ |
| $0$ | $1$ | $1$ |
| $1$ | $0$ | $1$ |
| $1$ | $1$ | $1$ |

<br></br>
The **AND** operation is also binary, but only results in a **True** statement if both of the variables are **True**. We express this operation as $(b \land a)$:
| $b$ | $a$ | $b \land a$ |
| :-: | :-: | :-: |
| $0$ | $0$ | $0$ |
| $0$ | $1$ | $0$ |
| $1$ | $0$ | $0$ |
| $1$ | $1$ | $1$ |
<br></br>

We can then combine these three basic operations to generate any possible logical statement between an arbitrary number of bits. For example, the 3-bit truth table below describes a statement that is **True** only when the majority of the bits are **False**.

| $c$ | $b$ | $a$ | $(\bar{c} \land \bar{b}) \lor (\bar{c} \land \bar{a}) \lor (\bar{b} \land \bar{a}) $ |
| :-: | :-: | :-: | :-: |
| $0$ | $0$ | $0$ | $1$ |
| $0$ | $0$ | $1$ | $1$ |
| $0$ | $1$ | $0$ | $1$ |
| $0$ | $1$ | $1$ | $0$ |
| $1$ | $0$ | $0$ | $1$ |
| $1$ | $0$ | $1$ | $0$ |
| $1$ | $1$ | $0$ | $0$ |
| $1$ | $1$ | $1$ | $0$ |
<br></br>

It is worth point out that Boolean statements can always be simplified to use the least amount of literals as possible by using a combination of [Boolean Algebra laws](https://en.wikipedia.org/wiki/Boolean_algebra#Laws) and [De Morgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws), but we will not go into these details herein.
<br></br>

A very important composite Boolean operation that we will use aplenty throughout the textbook is the [Exclusive Disjunction](https://en.wikipedia.org/wiki/Exclusive_or), or Exclusive OR (**XOR**). The **XOR** operation is **True** if only one of the two variables is **True**.

| $b$ | $a$ | $b \oplus a$ |
| :-: | :-: | :-: |
| $0$ | $0$ | $0$ |
| $0$ | $1$ | $1$ |
| $1$ | $0$ | $1$ |
| $1$ | $1$ | $0$ |
<br></br>

The  **XOR** is a composite operation is because it can be expressed in terms of **NOT**, **OR** and **AND** operations as follows:

$$ b \oplus a = (b \land \bar{a}) \lor (\bar{b} \land a) .$$

However, another way to interpret the **XOR** operation (when using binary values) is that it is equivalent to the addition operation modulo 2:

$$ b \oplus a = (b + a) \text{ mod } 2. $$
<br></br>

When dealing with binary numbers of more than one bit, we will use Boolean logic operators to act bitwise. So, for example, if we have two 3-bit numbers $a = a_2 a_1 a_0$ and $b = b_2 b_1 b_0$, the operation $a \oplus b$ represents the bitwise **XOR** between the two:

$$
\begin{aligned}
c &= a \oplus b
\\
c_2 c_1 c_0 &= (a_2 a_1 a_0) \oplus (b_2 b_1 b_0) ,
\end{aligned}
$$

where $c_i = a_i \oplus b_i$.

The table below summarizes the bitwise operator symbols used in Python, followed by a some examples:

| Operator | Python |
| :-: | :-: |
| AND | `&` |
| OR | `|` |
| XOR | `^` |

$ $

In [13]:
a = 0b11010
b = 0b10110

c_and = np.binary_repr(a & b, 5)  # Bitwise AND: 11010 & 10110 = 10010
c_or = np.binary_repr(a | b, 5)   # Bitwise OR:  11010 | 10110 = 11110
c_xor = np.binary_repr(a ^ b, 5)  # Bitwise XOR: 11010 ^ 10110 = 01100

print(f'a AND b = {c_and}') 
print(f'a OR  b = {c_or}')
print(f'a XOR b = {c_xor}')

a AND b = 10010
a OR  b = 11110
a XOR b = 01100


Something very important to note here is that the Python logical operators `and` and `or` are **not** equivalent to the bitwise operators `&` and `^` when applied to binary integers of more than one bit. Understanding how these operators act on binary integers is irrelevant to the current discussion, but if you're curious about it, you can check this [Stack Overflow post](https://stackoverflow.com/questions/49658308/how-does-the-logical-and-operator-work-with-integers).

## 3. Digital Circuits

Digital circuits are a diagrammatic representation of Boolean logic operations that directly maps to a physical implementation using electronic hardware. Understanding circuits is of great interest to us because different computational problems can be mapped onto this model, facilitating their classification into different complexity classes by using [circuit complexity theory](https://en.wikipedia.org/wiki/Circuit_complexity). As we will see in later chapters, this approach is widely used to analyze the type of advantage quantum algorithms can offer over a classical approach.


Classical circuits are constructed by interconnecting fundamental [logic gates](https://en.wikipedia.org/wiki/Boolean_algebra#Digital_logic_gates) associated with the three basic Boolean operations: **NOT**, **OR**, **AND**. The figure below shows the gate symbols for these three operators (from left to right):

<img src="images\01_01_01_basic_logic_gates.png" align = "center" width="750"/>

Similarly, the **XOR** operator is so widely used that it has its own symbol, even though it can be constructed out of **NOT**, **OR** and **AND** gates:

<img src="images\01_01_02_xor.png" align = "center" width="180"/>

Another schematic simplification is to use a circle at the inputs or output of the binary gates to "absorb" a negation. For example, a **AND** / **OR** / **XOR** gate immediately followed by a **NOT** gate can be represented as:

<img src="images\01_01_03_negated_logic_gates.png" align = "center" width="790"/>

These three gates are common enough that receive the names **NAND** / **NOR** / **XNOR**.
<br></br>

Using only these fundamental gates, we can then compose digital circuits for more complicated Boolean operations. For example, let's say we want to construct a circuit that takes two bits $a$, $b$, and an input carry $c_{in}$, and adds them into an output $s$, and an output carry $c_{out}$. We can express this in the form of a truth table:

| $a$ | $b$ | $c_{in}$ | $s$ | $c_{out}$ |
| :-: | :-: | :-: | :-: | :-: |
| $0$ | $0$ | $0$ | $0$ | $0$ |
| $0$ | $0$ | $1$ | $1$ | $0$ |
| $0$ | $1$ | $0$ | $1$ | $0$ |
| $0$ | $1$ | $1$ | $0$ | $1$ |
| $1$ | $0$ | $0$ | $1$ | $0$ |
| $1$ | $0$ | $1$ | $0$ | $1$ |
| $1$ | $1$ | $0$ | $0$ | $1$ |
| $1$ | $1$ | $1$ | $1$ | $1$ |
<br></br>

It is not difficult to show that $s = a \oplus b \oplus c_{in}$, which is expected since the **XOR** operation is equivalent to  addition modulo 2. Finding the expression for $c_{out}$ is slightly more involved since it requires synthesizing a Boolean expression by using methods like the [sum of products conversion](https://en.wikipedia.org/wiki/Disjunctive_normal_form#Conversion_to_DNF) or [Karnaugh maps](https://en.wikipedia.org/wiki/Karnaugh_map), which we will not cover here. It suffices to say that it can be shown that $c_{out} = ((a \oplus b) \land c_{in}) \lor (a \land b) $, which results in the circuit shown below:


<img src="images\01_01_04_adder.png" align = "center" width="420"/>

A concept that we have not yet discussed but is commonly utilized in circuit complexity analysis is that of the **COPY** or **FAN-OUT** operation. In the circuit above, we see that, for example, the variable $a$ is used as an input on two separate gates. This variable was copied (or fanned-out) when a single wire turned into two separate ones:

<img src="images\01_01_05_fanout.png" align = "center" width="80"/>

From a schematic standpoint, this might seem trivial considering that this will only require connecting an additional wire. However, from the point of complexity analysis this is an important operation because, as we will see later, there is no equivalent to the **COPY** operation in quantum circuits.
<br></br>

The key insight of being able to implement more complex Boolean circuits out of fundamental logic gates is that we can follow this process to construct all the building blocks of modern classical computing systems. Using only these gates, we can not only develop all the [combinatorial logic](https://en.wikipedia.org/wiki/Combinational_logic) used to build, for example, the core [arithmetic logic unit (ALU)](https://en.wikipedia.org/wiki/Arithmetic_logic_unit) in a [central processing unit (CPU)](https://en.wikipedia.org/wiki/Central_processing_unit), but also all [sequential logic](https://en.wikipedia.org/wiki/Sequential_logic) (latches, flip-flops, registers) required to store and parse its data.
<br></br>

Lastly, it is worth mentioning that, at a physical level, digital logic gates are not fundamental but are rather implemented using transistors. At this point, we will not go into any details on how these devices work, but hopefully in the future this section will be expanded to include some of the basics. This is important if we want to understand how quantum processors differ from classical devices at the hardware level.
<br></br>

In the next chapter, we will continue to build our understanding towards quantum computing. An important step in this process is to introduce a different model of computation known as [reversible computing](https://en.wikipedia.org/wiki/Reversible_computing). This is an important step in this process because the physical implementation of quantum hardware relies on quantum mechanics, which in principle is a reversible theory and therefore its model of computation should be reversible too.